proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1294 @@
+
+   In order to facilitate writing code that is forward-compatible with
+   ``RequiredTypeArguments``, introduce a new warning to ``-Wcompat``: ``-Wterm-variable-capture``.
+   This warning will notify users when implicit quantification occurs that
+   would stop working under ``RequiredTypeArguments``.
+
+Type checking
+~~~~~~~~~~~~~
+
+9. Introduce an additional typing rule that transforms term arguments into type
+   arguments::
+
+     rho = t2t(e)
+     G |- (forall a -> sigma); type rho, pis  ~>  Theta; phis; rho_r
+     ----------------------------------------------------------------  T2T
+     G |- (forall a -> sigma);        e, pis  ~>  Theta; phis; rho_r
+
+   See "T2T-Mapping" below for an informal definition of ``t2t``.
+
+   In other words, given ``f :: forall a -> t``, the ``x`` in ``f x`` is
+   parsed and renamed as a term, but then mapped to a type.
+
+
+10. Any uses of terms in types are ill-typed:
+    ::
+
+      a = 42; f :: Proxy a -> Proxy b   -- invalid occurrence of "a" in a type-level position
+
+    Any uses of types in terms that do not undergo the T2T transformation are also ill-typed::
+    ::
+
+      f _ = Int                         -- invalid occurrence of "Int" in a term-level position
+
+11. When in the checking mode of bidirectional type checking (e.g. in a function
+    binding with an explicit type signature), allow a pattern to bind type
+    variables in the term namespace, such as ``x`` here::
+
+      f :: forall a -> ...
+      f x = ...
+
+    The ``x`` identifier is bound in the term namespace, but stands for an
+    erased, ``forall``-bound type variable.
+
+    Just as with patterns that use the ``type`` herald explicitly, we permit
+    only variables and wildcards in such positions::
+
+      f :: forall a -> ...
+      f x   = ...               -- OK
+      f _   = ...               -- OK
+      f Int = ...               -- illegal to match on a type
+
+    We could, but we do not relax this requirement even when the type is statically known::
+
+      data T a where
+        Typed :: forall a -> a -> T a
+
+      g :: T Int -> ...
+      g (Typed Int x) = ...   -- Reasonable, but no.
+
+    This is a conservative decision that can be revised in a later proposal.
+
+Namespaces vs Semantics
+~~~~~~~~~~~~~~~~~~~~~~~
+
+12. With the proposed changes, the namespace of an identifier is no longer tied
+    to whether it stands for a type variable or a term variable.
+
+    Before this proposal, all term variables (retained, values, runtime) used
+    names from the term namespace, and all type variables (erased, types,
+    compile-time) used names from the type namespace. This is no longer the
+    case.
+
+T2T-Mapping
+~~~~~~~~~~~
+
+T2T (term-to-type) is a mapping from expressions to types that operates on a
+resolved syntax tree and is invoked by the ``T2T`` typing rule.
+
+The T2T mapping is partial: it succeeds on expressions that are within the
```

Comment:
Can we not write this more formally?
```
t2t[ x ]                = x
t2t[ e1 e2 ]         = t2t[e1]  t2t[e2]
t2t[ (e1, .., en) ]   =  '(  t2t[e1], ..., t2t[en] )
t2t[ e1 -> e2 ]    = t2t[e1]  ->  t2t[e2]
```
etc

That would be more systematic than a wall of bullets, wouldn't it?  We are functional programmers, after all.

Reply:
I started rewriting it as you suggest, but I think it’s not as helpful and makes it hard to distinguish Haskell syntax and meta-notation. For instance, the rule for list literals becomes

```
t2t[ [e₀, e₁, ...] ] = [t2t[e₀], t2t[e₁], ...]
```

It gets confusing because square brackets mean two different things here (and if we used round parens, the same issue would arise for tuples).

> We are functional programmers, after all.

The implementation will surely be a function :-)

Reply:
If delimiter ambiguity is the problem, you can use some unicode delimiters. Popular ones for such an operations would be ⟦ ⟧ or ⟪ ⟫.